File: tests/math/Rotor4D.test.js
/**
 * Rotor4D Unit Tests
 */

import { describe, it, expect } from 'vitest';
import { Rotor4D } from '../../src/math/Rotor4D.js';
import { Vec4 } from '../../src/math/Vec4.js';

describe('Rotor4D', () => {
    describe('constructor and identity', () => {
        it('creates identity rotor by default', () => {
            const r = new Rotor4D();
            expect(r.s).toBe(1);
            expect(r.xy).toBe(0);
            expect(r.xyzw).toBe(0);
        });

        it('creates identity with static method', () => {
            const r = Rotor4D.identity();
            expect(r.s).toBe(1);
            expect(r.norm()).toBeCloseTo(1);
        });

        it('has 8 components', () => {
            const r = new Rotor4D(1, 2, 3, 4, 5, 6, 7, 8);
            expect(r.s).toBe(1);
            expect(r.xy).toBe(2);
            expect(r.xz).toBe(3);
            expect(r.yz).toBe(4);
            expect(r.xw).toBe(5);
            expect(r.yw).toBe(6);
            expect(r.zw).toBe(7);
            expect(r.xyzw).toBe(8);
        });
    });

    describe('normalization', () => {
        it('normalizes to unit length', () => {
            const r = new Rotor4D(1, 1, 1, 1, 1, 1, 1, 1);
            const n = r.normalize();
            expect(n.norm()).toBeCloseTo(1);
        });

        it('preserves identity when normalizing', () => {
            const r = Rotor4D.identity();
            const n = r.normalize();
            expect(n.norm()).toBeCloseTo(1);
            expect(n.s).toBeCloseTo(1);
        });

        it('handles zero rotor', () => {
            const r = new Rotor4D(0, 0, 0, 0, 0, 0, 0, 0);
            const n = r.normalize();
            expect(n.isIdentity()).toBe(true);
        });
    });

    describe('plane rotations', () => {
        it('creates XY rotation', () => {
            const r = Rotor4D.fromPlaneAngle('XY', Math.PI / 2);
            expect(r.norm()).toBeCloseTo(1);
        });

        it('creates all 6 plane rotations', () => {
            const planes = ['XY', 'XZ', 'YZ', 'XW', 'YW', 'ZW'];
            for (const plane of planes) {
                const r = Rotor4D.fromPlaneAngle(plane, Math.PI / 4);
                expect(r.norm()).toBeCloseTo(1);
            }
        });

        it('identity rotation returns identity rotor', () => {
            const r = Rotor4D.fromPlaneAngle('XY', 0);
            expect(r.s).toBeCloseTo(1);
        });

        it('throws for invalid plane', () => {
            expect(() => Rotor4D.fromPlaneAngle('AB', Math.PI)).toThrow();
        });
    });

    describe('rotation application', () => {
        it('identity does not change vector', () => {
            const r = Rotor4D.identity();
            const v = new Vec4(1, 2, 3, 4);
            const rotated = r.rotate(v);
            expect(rotated.equals(v, 1e-5)).toBe(true);
        });

        it('XY rotation rotates in XY plane', () => {
            const r = Rotor4D.fromPlaneAngle('XY', Math.PI / 2);
            const v = Vec4.unitX();
            const rotated = r.rotate(v);
            // X should become Y
            expect(rotated.y).toBeCloseTo(1, 4);
            expect(rotated.x).toBeCloseTo(0, 4);
            expect(rotated.z).toBeCloseTo(0, 4);
            expect(rotated.w).toBeCloseTo(0, 4);
        });

        it('XW rotation moves X into W', () => {
            const r = Rotor4D.fromPlaneAngle('XW', Math.PI / 2);
            const v = Vec4.unitX();
            const rotated = r.rotate(v);
            // X should become W
            expect(rotated.w).toBeCloseTo(1, 4);
            expect(rotated.x).toBeCloseTo(0, 4);
        });

        it('preserves vector length', () => {
            const r = Rotor4D.fromPlaneAngle('XW', 1.234);
            const v = new Vec4(1, 2, 3, 4);
            const rotated = r.rotate(v);
            expect(rotated.length()).toBeCloseTo(v.length(), 4);
        });

        it('full rotation returns to start', () => {
            const r = Rotor4D.fromPlaneAngle('YZ', Math.PI * 2);
            const v = new Vec4(1, 2, 3, 4);
            const rotated = r.rotate(v);
            expect(rotated.equals(v, 1e-4)).toBe(true);
        });
    });

    describe('composition', () => {
        it('composes rotations', () => {
            const r1 = Rotor4D.fromPlaneAngle('XY', Math.PI / 4);
            const r2 = Rotor4D.fromPlaneAngle('XY', Math.PI / 4);
            const combined = r1.multiply(r2);
            // Should be equivalent to XY rotation by PI/2
            expect(combined.norm()).toBeCloseTo(1);
        });

        it('identity * r = r', () => {
            const id = Rotor4D.identity();
            const r = Rotor4D.fromPlaneAngle('ZW', 1.5);
            const result = id.multiply(r);
            expect(result.equals(r)).toBe(true);
        });

        it('r * r.inverse() = identity', () => {
            const r = Rotor4D.fromPlaneAngle('XW', 0.7);
            const inv = r.inverse();
            const result = r.multiply(inv);
            expect(result.isIdentity()).toBe(true);
        });
    });

    describe('matrix conversion', () => {
        it('converts to 4x4 matrix', () => {
            const r = Rotor4D.fromPlaneAngle('XY', Math.PI / 4);
            const m = r.toMatrix();
            expect(m).toBeInstanceOf(Float32Array);
            expect(m.length).toBe(16);
        });

        it('identity rotor gives identity matrix', () => {
            const r = Rotor4D.identity();
            const m = r.toMatrix();
            // Check diagonal
            expect(m[0]).toBeCloseTo(1);
            expect(m[5]).toBeCloseTo(1);
            expect(m[10]).toBeCloseTo(1);
            expect(m[15]).toBeCloseTo(1);
            // Check some off-diagonal
            expect(m[1]).toBeCloseTo(0);
            expect(m[4]).toBeCloseTo(0);
        });
    });

    describe('interpolation', () => {
        it('slerps between rotors', () => {
            const r1 = Rotor4D.fromPlaneAngle('XW', 0);
            const r2 = Rotor4D.fromPlaneAngle('XW', Math.PI);
            const mid = r1.slerp(r2, 0.5);
            expect(mid.norm()).toBeCloseTo(1);
        });

        it('slerp at t=0 returns start', () => {
            const r1 = Rotor4D.fromPlaneAngle('YZ', 0.5);
            const r2 = Rotor4D.fromPlaneAngle('YZ', 1.5);
            const result = r1.slerp(r2, 0);
            expect(result.equals(r1)).toBe(true);
        });

        it('slerp at t=1 returns end', () => {
            const r1 = Rotor4D.fromPlaneAngle('YZ', 0.5);
            const r2 = Rotor4D.fromPlaneAngle('YZ', 1.5);
            const result = r1.slerp(r2, 1);
            expect(result.equals(r2)).toBe(true);
        });
    });

    describe('numerical stability', () => {
        it('stays normalized after many multiplications', () => {
            let r = Rotor4D.fromPlaneAngle('XW', 0.01);
            for (let i = 0; i < 100; i++) {
                r = r.multiply(Rotor4D.fromPlaneAngle('YZ', 0.01));
                r.normalizeInPlace();
            }
            expect(r.norm()).toBeCloseTo(1, 5);
        });

        it('rotation preserves vector length over many iterations', () => {
            const r = Rotor4D.fromPlaneAngle('ZW', 0.1);
            let v = new Vec4(1, 2, 3, 4);
            const originalLength = v.length();

            for (let i = 0; i < 100; i++) {
                v = r.rotate(v);
            }

            expect(v.length()).toBeCloseTo(originalLength, 4);
        });
    });

    describe('fromEuler6', () => {
        it('creates rotor from 6 angles', () => {
            const r = Rotor4D.fromEuler6({
                xy: 0.1, xz: 0.2, yz: 0.3,
                xw: 0.4, yw: 0.5, zw: 0.6
            });
            expect(r.norm()).toBeCloseTo(1);
        });

        it('zero angles gives identity', () => {
            const r = Rotor4D.fromEuler6({});
            expect(r.isIdentity()).toBe(true);
        });
    });
});

File: cpp/tests/Rotor4D_test.cpp
/**
 * Rotor4D_test.cpp - Unit tests for Rotor4D 4D rotation class
 *
 * Tests identity construction, plane-angle rotors, Euler6 composition,
 * sandwich product rotation, normalization, and reverse (conjugate).
 */

#include <gtest/gtest.h>
#include "math/Rotor4D.hpp"
#include "math/Vec4.hpp"
#include <cmath>
#include <numbers>

using namespace vib3;

static constexpr float kEpsilon = 1e-4f;
static constexpr float kPi = std::numbers::pi_v<float>;
static constexpr float kHalfPi = kPi / 2.0f;

// ============================================================================
// Identity Rotor
// ============================================================================

TEST(Rotor4D, DefaultConstructorIsIdentity) {
    Rotor4D r;
    EXPECT_FLOAT_EQ(r.s, 1.0f);
    EXPECT_FLOAT_EQ(r.xy, 0.0f);
    EXPECT_FLOAT_EQ(r.xz, 0.0f);
    EXPECT_FLOAT_EQ(r.yz, 0.0f);
    EXPECT_FLOAT_EQ(r.xw, 0.0f);
    EXPECT_FLOAT_EQ(r.yw, 0.0f);
    EXPECT_FLOAT_EQ(r.zw, 0.0f);
    EXPECT_FLOAT_EQ(r.xyzw, 0.0f);
}

TEST(Rotor4D, IdentityFactory) {
    Rotor4D r = Rotor4D::identity();
    EXPECT_FLOAT_EQ(r.s, 1.0f);
    EXPECT_FLOAT_EQ(r.xy, 0.0f);
    EXPECT_FLOAT_EQ(r.xz, 0.0f);
    EXPECT_FLOAT_EQ(r.yz, 0.0f);
    EXPECT_FLOAT_EQ(r.xw, 0.0f);
    EXPECT_FLOAT_EQ(r.yw, 0.0f);
    EXPECT_FLOAT_EQ(r.zw, 0.0f);
    EXPECT_FLOAT_EQ(r.xyzw, 0.0f);
}

TEST(Rotor4D, IdentityIsNormalized) {
    Rotor4D r = Rotor4D::identity();
    EXPECT_TRUE(r.isNormalized());
    EXPECT_NEAR(r.magnitude(), 1.0f, kEpsilon);
}

TEST(Rotor4D, IdentityRotationDoesNothing) {
    Rotor4D r = Rotor4D::identity();
    Vec4 v(1.0f, 2.0f, 3.0f, 4.0f);
    Vec4 rotated = r.rotate(v);
    EXPECT_NEAR(rotated.x, v.x, kEpsilon);
    EXPECT_NEAR(rotated.y, v.y, kEpsilon);
    EXPECT_NEAR(rotated.z, v.z, kEpsilon);
    EXPECT_NEAR(rotated.w, v.w, kEpsilon);
}

// ============================================================================
// fromPlaneAngle
// ============================================================================

TEST(Rotor4D, FromPlaneAngleZeroIsIdentity) {
    Rotor4D r = Rotor4D::fromPlaneAngle(RotationPlane::XY, 0.0f);
    EXPECT_NEAR(r.s, 1.0f, kEpsilon);
    EXPECT_NEAR(r.magnitude(), 1.0f, kEpsilon);
}

TEST(Rotor4D, FromPlaneAngleXY_HalfTurn) {
    // A rotor encoding angle theta has s = cos(theta/2), bivector = -sin(theta/2)
    Rotor4D r = Rotor4D::fromPlaneAngle(RotationPlane::XY, kPi);
    EXPECT_NEAR(r.s, std::cos(kPi / 2.0f), kEpsilon);
    EXPECT_TRUE(r.isNormalized());
}

TEST(Rotor4D, FromPlaneAngleXY_RotatesXtoY) {
    // Rotating unitX by pi/2 in the XY plane should give unitY
    Rotor4D r = Rotor4D::fromPlaneAngle(RotationPlane::XY, kHalfPi);
    Vec4 result = r.rotate(Vec4::unitX());
    EXPECT_NEAR(result.x, 0.0f, kEpsilon);
    EXPECT_NEAR(result.y, 1.0f, kEpsilon);
    EXPECT_NEAR(result.z, 0.0f, kEpsilon);
    EXPECT_NEAR(result.w, 0.0f, kEpsilon);
}

TEST(Rotor4D, FromPlaneAngleXY_RotatesYtoNegX) {
    // Rotating unitY by pi/2 in XY plane should give -unitX
    Rotor4D r = Rotor4D::fromPlaneAngle(RotationPlane::XY, kHalfPi);
    Vec4 result = r.rotate(Vec4::unitY());
    EXPECT_NEAR(result.x, -1.0f, kEpsilon);
    EXPECT_NEAR(result.y, 0.0f, kEpsilon);
    EXPECT_NEAR(result.z, 0.0f, kEpsilon);
    EXPECT_NEAR(result.w, 0.0f, kEpsilon);
}

TEST(Rotor4D, FromPlaneAngleXZ_RotatesXtoZ) {
    Rotor4D r = Rotor4D::fromPlaneAngle(RotationPlane::XZ, kHalfPi);
    Vec4 result = r.rotate(Vec4::unitX());
    EXPECT_NEAR(result.x, 0.0f, kEpsilon);
    EXPECT_NEAR(result.y, 0.0f, kEpsilon);
    EXPECT_NEAR(result.z, 1.0f, kEpsilon);
    EXPECT_NEAR(result.w, 0.0f, kEpsilon);
}

TEST(Rotor4D, FromPlaneAngleYZ_RotatesYtoZ) {
    Rotor4D r = Rotor4D::fromPlaneAngle(RotationPlane::YZ, kHalfPi);
    Vec4 result = r.rotate(Vec4::unitY());
    EXPECT_NEAR(result.x, 0.0f, kEpsilon);
    EXPECT_NEAR(result.y, 0.0f, kEpsilon);
    EXPECT_NEAR(result.z, 1.0f, kEpsilon);
    EXPECT_NEAR(result.w, 0.0f, kEpsilon);
}

TEST(Rotor4D, FromPlaneAngleXW_RotatesXtoW) {
    // 4D rotation: XW plane, pi/2 should rotate X -> W
    Rotor4D r = Rotor4D::fromPlaneAngle(RotationPlane::XW, kHalfPi);
    Vec4 result = r.rotate(Vec4::unitX());
    EXPECT_NEAR(result.x, 0.0f, kEpsilon);
    EXPECT_NEAR(result.y, 0.0f, kEpsilon);
    EXPECT_NEAR(result.z, 0.0f, kEpsilon);
    EXPECT_NEAR(result.w, 1.0f, kEpsilon);
}

TEST(Rotor4D, FromPlaneAngleYW_RotatesYtoW) {
    Rotor4D r = Rotor4D::fromPlaneAngle(RotationPlane::YW, kHalfPi);
    Vec4 result = r.rotate(Vec4::unitY());
    EXPECT_NEAR(result.x, 0.0f, kEpsilon);
    EXPECT_NEAR(result.y, 0.0f, kEpsilon);
    EXPECT_NEAR(result.z, 0.0f, kEpsilon);
    EXPECT_NEAR(result.w, 1.0f, kEpsilon);
}

TEST(Rotor4D, FromPlaneAngleZW_RotatesZtoW) {
    Rotor4D r = Rotor4D::fromPlaneAngle(RotationPlane::ZW, kHalfPi);
    Vec4 result = r.rotate(Vec4::unitZ());
    EXPECT_NEAR(result.x, 0.0f, kEpsilon);
    EXPECT_NEAR(result.y, 0.0f, kEpsilon);
    EXPECT_NEAR(result.z, 0.0f, kEpsilon);
    EXPECT_NEAR(result.w, 1.0f, kEpsilon);
}

TEST(Rotor4D, PlaneRotationPreservesOrthogonalAxes) {
    // Rotation in XY plane should not affect Z or W components
    Rotor4D r = Rotor4D::fromPlaneAngle(RotationPlane::XY, kHalfPi);
    Vec4 zResult = r.rotate(Vec4::unitZ());
    EXPECT_NEAR(zResult.x, 0.0f, kEpsilon);
    EXPECT_NEAR(zResult.y, 0.0f, kEpsilon);
    EXPECT_NEAR(zResult.z, 1.0f, kEpsilon);
    EXPECT_NEAR(zResult.w, 0.0f, kEpsilon);

    Vec4 wResult = r.rotate(Vec4::unitW());
    EXPECT_NEAR(wResult.x, 0.0f, kEpsilon);
    EXPECT_NEAR(wResult.y, 0.0f, kEpsilon);
    EXPECT_NEAR(wResult.z, 0.0f, kEpsilon);
    EXPECT_NEAR(wResult.w, 1.0f, kEpsilon);
}

// ============================================================================
// fromEuler6
// ============================================================================

TEST(Rotor4D, FromEuler6AllZerosIsIdentity) {
    Rotor4D r = Rotor4D::fromEuler6(0.0f, 0.0f, 0.0f, 0.0f, 0.0f, 0.0f);
    EXPECT_NEAR(r.s, 1.0f, kEpsilon);
    EXPECT_TRUE(r.isNormalized());

    // Verify it acts as identity
    Vec4 v(1.0f, 2.0f, 3.0f, 4.0f);
    Vec4 result = r.rotate(v);
    EXPECT_NEAR(result.x, v.x, kEpsilon);
    EXPECT_NEAR(result.y, v.y, kEpsilon);
    EXPECT_NEAR(result.z, v.z, kEpsilon);
    EXPECT_NEAR(result.w, v.w, kEpsilon);
}

TEST(Rotor4D, FromEuler6ArrayOverload) {
    std::array<float, 6> angles = {0.0f, 0.0f, 0.0f, 0.0f, 0.0f, 0.0f};
    Rotor4D r = Rotor4D::fromEuler6(angles);
    EXPECT_NEAR(r.s, 1.0f, kEpsilon);
    EXPECT_TRUE(r.isNormalized());
}

TEST(Rotor4D, FromEuler6SinglePlaneMatchesFromPlaneAngle) {
    float angle = 0.7f;
    Rotor4D fromEuler = Rotor4D::fromEuler6(angle, 0.0f, 0.0f, 0.0f, 0.0f, 0.0f);
    Rotor4D fromPlane = Rotor4D::fromPlaneAngle(RotationPlane::XY, angle);

    // They should produce the same rotation on any vector
    Vec4 test(1.0f, 1.0f, 1.0f, 1.0f);
    Vec4 r1 = fromEuler.rotate(test);
    Vec4 r2 = fromPlane.rotate(test);
    EXPECT_NEAR(r1.x, r2.x, kEpsilon);
    EXPECT_NEAR(r1.y, r2.y, kEpsilon);
    EXPECT_NEAR(r1.z, r2.z, kEpsilon);
    EXPECT_NEAR(r1.w, r2.w, kEpsilon);
}

TEST(Rotor4D, FromEuler6ProducesNormalizedRotor) {
    Rotor4D r = Rotor4D::fromEuler6(0.5f, 0.3f, 0.2f, 0.1f, 0.4f, 0.6f);
    // The composed rotor should still be approximately unit
    EXPECT_NEAR(r.magnitude(), 1.0f, 0.01f);
}

// ============================================================================
// Normalization
// ============================================================================

TEST(Rotor4D, NormalizeMaintainsUnit) {
    Rotor4D r = Rotor4D::fromPlaneAngle(RotationPlane::XY, 1.0f);
    r.normalize();
    EXPECT_NEAR(r.magnitude(), 1.0f, kEpsilon);
    EXPECT_TRUE(r.isNormalized());
}

TEST(Rotor4D, NormalizedReturnsUnitRotor) {
    // Manually create a non-unit rotor
    Rotor4D r(2.0f, 0.0f, 0.0f, 0.0f, 0.0f, 0.0f, 0.0f, 0.0f);
    Rotor4D n = r.normalized();
    EXPECT_NEAR(n.magnitude(), 1.0f, kEpsilon);
    EXPECT_NEAR(n.s, 1.0f, kEpsilon);
}

TEST(Rotor4D, MagnitudeSquared) {
    Rotor4D r = Rotor4D::identity();
    EXPECT_NEAR(r.magnitudeSquared(), 1.0f, kEpsilon);
}

// ============================================================================
// Reverse (Conjugate)
// ============================================================================

TEST(Rotor4D, ReverseNegatesBivectors) {
    Rotor4D r(1.0f, 0.1f, 0.2f, 0.3f, 0.4f, 0.5f, 0.6f, 0.7f);
    Rotor4D rev = r.reverse();

    // Scalar stays the same
    EXPECT_FLOAT_EQ(rev.s, r.s);
    // Bivectors are negated
    EXPECT_FLOAT_EQ(rev.xy, -r.xy);
    EXPECT_FLOAT_EQ(rev.xz, -r.xz);
    EXPECT_FLOAT_EQ(rev.yz, -r.yz);
    EXPECT_FLOAT_EQ(rev.xw, -r.xw);
    EXPECT_FLOAT_EQ(rev.yw, -r.yw);
    EXPECT_FLOAT_EQ(rev.zw, -r.zw);
    // Pseudoscalar: grade-4 element, reverse keeps it the same
    EXPECT_FLOAT_EQ(rev.xyzw, r.xyzw);
}

TEST(Rotor4D, ReverseOfIdentityIsIdentity) {
    Rotor4D r = Rotor4D::identity();
    Rotor4D rev = r.reverse();
    EXPECT_FLOAT_EQ(rev.s, 1.0f);
    EXPECT_FLOAT_EQ(rev.xy, 0.0f);
    EXPECT_FLOAT_EQ(rev.xz, 0.0f);
    EXPECT_FLOAT_EQ(rev.yz, 0.0f);
}

TEST(Rotor4D, RotorTimesReverseGivesIdentityRotation) {
    // R * R_reverse should produce identity rotation
    Rotor4D r = Rotor4D::fromPlaneAngle(RotationPlane::XY, 1.0f);
    Rotor4D product = r * r.reverse();

    // The result should act as identity on any vector
    Vec4 v(1.0f, 2.0f, 3.0f, 4.0f);
    Vec4 result = product.rotate(v);
    EXPECT_NEAR(result.x, v.x, kEpsilon);
    EXPECT_NEAR(result.y, v.y, kEpsilon);
    EXPECT_NEAR(result.z, v.z, kEpsilon);
    EXPECT_NEAR(result.w, v.w, kEpsilon);
}

// ============================================================================
// Rotor Composition
// ============================================================================

TEST(Rotor4D, IdentityTimesRotorIsRotor) {
    Rotor4D id = Rotor4D::identity();
    Rotor4D r = Rotor4D::fromPlaneAngle(RotationPlane::XZ, 0.5f);
    Rotor4D product = id * r;

    Vec4 test(1.0f, 0.0f, 0.0f, 0.0f);
    Vec4 r1 = product.rotate(test);
    Vec4 r2 = r.rotate(test);
    EXPECT_NEAR(r1.x, r2.x, kEpsilon);
    EXPECT_NEAR(r1.y, r2.y, kEpsilon);
    EXPECT_NEAR(r1.z, r2.z, kEpsilon);
    EXPECT_NEAR(r1.w, r2.w, kEpsilon);
}

TEST(Rotor4D, TwoHalfTurnsMakeFullTurn) {
    // Two 90-degree rotations = 180-degree rotation
    Rotor4D half = Rotor4D::fromPlaneAngle(RotationPlane::XY, kHalfPi);
    Rotor4D full = half * half;

    // unitX rotated by pi in XY plane should become -unitX
    Vec4 result = full.rotate(Vec4::unitX());
    EXPECT_NEAR(result.x, -1.0f, kEpsilon);
    EXPECT_NEAR(result.y, 0.0f, kEpsilon);
    EXPECT_NEAR(result.z, 0.0f, kEpsilon);
    EXPECT_NEAR(result.w, 0.0f, kEpsilon);
}

// ============================================================================
// Rotation preserves length
// ============================================================================

TEST(Rotor4D, RotationPreservesLength) {
    Rotor4D r = Rotor4D::fromEuler6(0.5f, 0.3f, 0.7f, 0.2f, 0.4f, 0.1f);
    Vec4 v(1.0f, 2.0f, 3.0f, 4.0f);
    Vec4 rotated = r.rotate(v);
    EXPECT_NEAR(rotated.length(), v.length(), kEpsilon);
}

TEST(Rotor4D, RotationPreservesZeroVector) {
    Rotor4D r = Rotor4D::fromPlaneAngle(RotationPlane::XW, 1.0f);
    Vec4 result = r.rotate(Vec4::zero());
    EXPECT_NEAR(result.x, 0.0f, kEpsilon);
    EXPECT_NEAR(result.y, 0.0f, kEpsilon);
    EXPECT_NEAR(result.z, 0.0f, kEpsilon);
    EXPECT_NEAR(result.w, 0.0f, kEpsilon);
}

// ============================================================================
// Inverse
// ============================================================================

TEST(Rotor4D, InverseUndoesRotation) {
    Rotor4D r = Rotor4D::fromPlaneAngle(RotationPlane::YZ, 0.8f);
    Rotor4D inv = r.inverse();
    Vec4 v(1.0f, 2.0f, 3.0f, 4.0f);
    Vec4 rotated = r.rotate(v);
    Vec4 unrotated = inv.rotate(rotated);
    EXPECT_NEAR(unrotated.x, v.x, kEpsilon);
    EXPECT_NEAR(unrotated.y, v.y, kEpsilon);
    EXPECT_NEAR(unrotated.z, v.z, kEpsilon);
    EXPECT_NEAR(unrotated.w, v.w, kEpsilon);
}

// ============================================================================
// Matrix Conversion
// ============================================================================

TEST(Rotor4D, IdentityRotorGivesIdentityMatrix) {
    Rotor4D r = Rotor4D::identity();
    Mat4x4 m = r.toMatrix();
    EXPECT_TRUE(m.isIdentity());
}

TEST(Rotor4D, MatrixRotationMatchesSandwichProduct) {
    Rotor4D r = Rotor4D::fromPlaneAngle(RotationPlane::XY, kHalfPi);
    Mat4x4 m = r.toMatrix();

    Vec4 v = Vec4::unitX();
    Vec4 rotorResult = r.rotate(v);
    Vec4 matResult = m * v;

    EXPECT_NEAR(rotorResult.x, matResult.x, kEpsilon);
    EXPECT_NEAR(rotorResult.y, matResult.y, kEpsilon);
    EXPECT_NEAR(rotorResult.z, matResult.z, kEpsilon);
    EXPECT_NEAR(rotorResult.w, matResult.w, kEpsilon);
}

// ============================================================================
// Interpolation
// ============================================================================

TEST(Rotor4D, SlerpEndpoints) {
    Rotor4D a = Rotor4D::identity();
    Rotor4D b = Rotor4D::fromPlaneAngle(RotationPlane::XY, kPi);

    // t=0 should give a
    Rotor4D s0 = a.slerp(b, 0.0f);
    Vec4 v0 = s0.rotate(Vec4::unitX());
    Vec4 va = a.rotate(Vec4::unitX());
    EXPECT_NEAR(v0.x, va.x, kEpsilon);
    EXPECT_NEAR(v0.y, va.y, kEpsilon);

    // t=1 should give b
    Rotor4D s1 = a.slerp(b, 1.0f);
    Vec4 v1 = s1.rotate(Vec4::unitX());
    Vec4 vb = b.rotate(Vec4::unitX());
    EXPECT_NEAR(v1.x, vb.x, kEpsilon);
    EXPECT_NEAR(v1.y, vb.y, kEpsilon);
}

// ============================================================================
// Comparison and toArray
// ============================================================================

TEST(Rotor4D, EqualityOperator) {
    Rotor4D a = Rotor4D::identity();
    Rotor4D b = Rotor4D::identity();
    EXPECT_TRUE(a == b);
}

TEST(Rotor4D, ToArrayRoundTrip) {
    Rotor4D r(1.0f, 0.1f, 0.2f, 0.3f, 0.4f, 0.5f, 0.6f, 0.7f);
    auto arr = r.toArray();
    EXPECT_EQ(arr.size(), 8u);
    EXPECT_FLOAT_EQ(arr[0], r.s);
    EXPECT_FLOAT_EQ(arr[1], r.xy);
    EXPECT_FLOAT_EQ(arr[2], r.xz);
    EXPECT_FLOAT_EQ(arr[3], r.yz);
    EXPECT_FLOAT_EQ(arr[4], r.xw);
    EXPECT_FLOAT_EQ(arr[5], r.yw);
    EXPECT_FLOAT_EQ(arr[6], r.zw);
    EXPECT_FLOAT_EQ(arr[7], r.xyzw);
}

// ============================================================================
// Dot product between rotors
// ============================================================================

TEST(Rotor4D, DotProductWithSelf) {
    Rotor4D r = Rotor4D::fromPlaneAngle(RotationPlane::XY, 0.5f);
    float d = r.dot(r);
    EXPECT_NEAR(d, r.magnitudeSquared(), kEpsilon);
}
